src/Reflex/PubSub.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE ExplicitForAll        #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE RecursiveDo           #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Reflex.PubSub where
+
+import Prelude                hiding (lookup)
+
+import Control.Monad.Reader
+import Data.Constraint.Extras
+import Data.Dependent.Map
+import Data.Dependent.Sum
+import Data.Functor.Identity
+
+import Reflex
+
+type PubSubT t tag m = EventWriterT t [DSum tag Identity] (ReaderT (Event t (DMap tag Identity)) m)
+
+-- | 'PubSub' enables you to broadcast events from anywhere in your reflex network to
+-- any location that is subscribed to the topic they are published to. Topics are best
+-- representated as constructors of some custom GADT which carries a phantom type to
+-- constrain the messages that can be published to it.
```

Comment:
Maybe add a note that this is consciously allowing spooky action-at-a-distance? (Not as badly as some other architectures, because you've tracked it with the type, but it's there.) This can be a fine and valid design choice, but ideally you want to `runPubSubT` over the minimal section of code that you want to talk to each other? (Like how it can be helpful to minimise the size of `rec` blocks.)

Reply:
Yeah, that's a good point. This can effectively recreate the "who wrote to this database table?" problem on your frontend, so you need to be aware of how that effects debuggability. Tracking with types helps some but it still doesn't tell you who published some event without a thorough grep. Also, since the problem it's meant to solve is handling long-range communication, trying to minimize it's scope kind of defeats the purpose, so that doesn't really work as a measure of whether it's the right thing to do.